### PR TITLE
add tracking of NotNullInfo for Match, Case, Try trees (fix #21380)

### LINKED ISSUE #21380: -Yexplicit-nulls fails to detect null pointer dereference
`-Yexplicit-nulls` fails to detect method call on a variable holding `null`.
This might be regression, as `scalac` v3.3.3 detects the error as expected.

## Compiler version
3.4.2

## Minimized code

```Scala
@main def test() = {
  var x: String | Null = null
  if (false) {
    x = ""

  } else {
    x = ""
  }
  try {
    x = ""
    throw new Exception()
  }
  catch {
    case e: Exception => {
      x = null
    }
  }
  x.replace("", "") 
}
```

## Output

The code compiles with `-Yexplicit-nulls`, but at runtime you get NPE

```scala
java.lang.NullPointerException
	at test$package$.test(test.scala:18)
	at test.main(test.scala:1)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at dotty.tools.runner.RichClassLoader$.run$extension$$anonfun$1(ScalaClassLoader.scala:36)
	at dotty.tools.runner.ScalaClassLoader$.asContext(ScalaClassLoader.scala:80)
	at dotty.tools.runner.RichClassLoader$.dotty$tools$runner$RichClassLoader$$$asContext$extension(ScalaClassLoader.scala:18)
	at dotty.tools.runner.RichClassLoader$.run$extension(ScalaClassLoader.scala:36)
	at dotty.tools.runner.CommonRunner.run(ObjectRunner.scala:23)
	at dotty.tools.runner.CommonRunner.run$(ObjectRunner.scala:13)
	at dotty.tools.runner.ObjectRunner$.run(ObjectRunner.scala:48)
	at dotty.tools.runner.CommonRunner.runAndCatch(ObjectRunner.scala:30)
	at dotty.tools.runner.CommonRunner.runAndCatch$(ObjectRunner.scala:13)
	at dotty.tools.runner.ObjectRunner$.runAndCatch(ObjectRunner.scala:48)
	at dotty.tools.MainGenericRunner$.run$1(MainGenericRunner.scala:215)
	at dotty.tools.MainGenericRunner$.process(MainGenericRunner.scala:270)
	at dotty.tools.MainGenericRunner$.main(MainGenericRunner.scala:281)
	at dotty.tools.MainGenericRunner.main(MainGenericRunner.scala)
```

## Expectation

The code should have been rejected with
```
-- [E008] Not Found Error: test.scala:18:4 ----------------------------------------------------------------------------------------------------------------------------------------------------
18 |  x.replace("", "")
   |  ^^^^^^^^^
   |  value replace is not a member of String | Null.
   |  Since explicit-nulls is enabled, the selection is rejected because
   |  String | Null could be null at runtime.
   |  If you want to select replace without checking for a null value,
   |  insert a .nn before .replace or import scala.language.unsafeNulls.
1 error found
```

### PR DESCRIPTION
**Pull request by olhotak**

### CODE REVIEW

**Comment by olhotak** [comment]
This would benefit from additional tests. @noti0na1 would you be able to add some?

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -2120,14 +2120,18 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
         case1
       }
       .asInstanceOf[List[CaseDef]]
-    assignType(cpy.Match(tree)(sel, cases1), sel, cases1).cast(pt)
+    var nni = sel.notNullInfo
+    if(cases1.nonEmpty) nni = nni.seq(cases1.map(_.notNullInfo).reduce(_.alt(_)))
```
Comment (som-snytt) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:2124]: Is there a style guide for `if then`? That avoids the faux pas of making if look like an application.

**Comment by noti0na1** [comment]
I fixed the NotNullInfos for Try and added some more tests.

We want the retracted NotNullInfo from the body of `Try` to type check the cases and finalizer, so we have to type `tree.expr` first using some trick.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 2]
```diff
@@ -2203,13 +2203,11 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
...
-      if (pt1.isValueType) // insert a cast if body does not conform to expected type if we disregard gadt bounds
+      if pt1.isValueType then // insert a cast if body does not conform to expected type if we disregard gadt bounds
         body1 = body1.ensureConforms(pt1)(using originalCtx)
-      val nni = pat1.notNullInfo.seq(
-        guard1.notNullInfoIf(false).alt(
-          guard1.notNullInfoIf(true).seq(body1.notNullInfo)
-        )
-      )
+      val nni = pat1.notNullInfo
+        .seq(guard1.notNullInfoIf(false).alt(guard1.notNullInfoIf(true)))
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:2209]: If the guard is false, the body does not execute. I don't think it's sound to .seq the body after an .alt that includes the case of the guard being false.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 3]
```diff
@@ -2314,16 +2312,25 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
       untpd.Block(makeCanThrow(capabilityProof), expr)
 
   def typedTry(tree: untpd.Try, pt: Type)(using Context): Try = {
+    // We want to type check tree.expr first to comput NotNullInfo, but `addCanThrowCapabilities`
+    // uses the types of patterns in `tree.cases` to determine the capabilities.
+    // Hence, we create a copy of cases with empty body and type check that first, then type check
+    // the rest of the tree in order.
+    val casesEmptyBody1 = tree.cases.mapconserve(cpy.CaseDef(_)(body = EmptyTree))
+    val casesEmptyBody2 = typedCases(casesEmptyBody1, EmptyTree, defn.ThrowableType, WildcardType)
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:2320]: Does this work if the pattern uses a variable whose nullability is changed by the try block (`expr`)? Won't we be typing such a variable in the pattern with the wrong type here?

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 4]
```diff
@@ -2314,16 +2312,25 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
...
       expr1 :: cases1
     }: @unchecked
-    val finalizer1 = typed(tree.finalizer, defn.UnitType)
     val cases2 = cases2x.asInstanceOf[List[CaseDef]]
-    val nni = expr2.notNullInfo.retractedInfo.seq(
-      cases2.map(_.notNullInfo.retractedInfo).fold(NotNullInfo.empty)(_.alt(_))
-    ).seq(finalizer1.notNullInfo)
+
+    var nni = expr2.notNullInfo.retractedInfo
+    if cases2.nonEmpty then nni = nni.seq(cases2.map(_.notNullInfo).reduce(_.alt(_)))
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:2331]: It's possible that none of the cases match (i.e. that no exception is thrown). So we either need an .alt with an empty NotNullInfo or we need to take the .retractedInfo of the cases (those two alternatives should be equivalent).

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 2 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala:2209]: I'm still working on the fix for typedCase:

```scala
def test3(i: Int) =
  var x: String | Null = null
  i match
    case 1 if x != null => ()
    case _ => x = " "
  x.trim() // ok
```

**Comment by noti0na1** [comment]
I may not have time to work on this PR until next week. Feel free to modify/remove my commit.

**Comment by noti0na1** [comment]
> `scalac -Ybest-effort tests/neg/i13864.scala`

It is tricky to fix the error: in the example, when the pat of the case is typed, the type argument is bind with a fresh name, so if we type check the cases twice, the `CanThrow` Capability would refer to a different name.

**Comment by jchyb** [comment]
@noti0na1 Feel free to add `i13864.scala` to `compiler/test/dotc/neg-best-effort-pickling.blacklist` (and feel free to use that exclusion list liberally in the future).
Best effort should never hamper any progress on the issues in the compiler (I should probably make that more clear in the message in the CI, sorry about that!). But thank you for looking at this either way, it should help me fix this later on (usually the fix would be something like avoiding pickling certain parts, so now I know where to look.)

**Review by olhotak — COMMENTED** [review]
LGTM now, but github requires a review from someone other than me. @noti0na1 can you approve and merge?

**Review by noti0na1 — APPROVED** [review]

**Pull request merged**